accepted/engine-host-functions.md

```diff
@@ -0,0 +1,211 @@
+
+For host functions that require contextual data (e.g. WASI), the `set_context` method will be used for associating context with a `Store` and it can later be retrieved via `caller.store().get_context()`.
+ 
+## Changes to `wasmtime::Linker`
+ 
+There will be no changes to the API surface of `Linker` to support the changes proposed in this RFC.
+ 
+However, the `Linker` implementation will be changed such that it will fallback to calling `Store::get_engine_function` when resolving imports for `Linker::instantiate` and `Linker::module`.
+ 
+This should allow for overriding engine function imports in the `Linker` if an import of the same name has already been defined in the `Engine`.
+ 
+## The `wasmtime::IntoEngineFunc` trait
+ 
+Much like `IntoFunc` that is used to easily define functions using `Func::wrap`, an `IntoEngineFunc` trait will be implemented for `Fn([Caller], P1, P2...) -> R` for up to sixteen parameters that implement `WasmTy` and the return type implements `WasmRet`.
+ 
+`IntoEngineFunc` will also be implemented for ``Fn(Caller, &[Val], &mut [Val]) -> Result<Trap> + `static`` so that any number of parameters and return values can be supported for engine functions.
+ 
+This trait, like `IntoFunc`, is intended as an internal implementation detail and will not be documented.
+ 
+## Changes to `wasmtime_wasi::Wasi`
+ 
+`Wasi` is a type generated from `wasmtime-wiggle`.
+ 
+This proposal adds the following methods to the generated `Wasi` type:
+ 
+```rust
+/// Adds WASI functions to the given `Engine`.
+pub fn add_to_engine<T>(engine: &Engine) -> Result<()>;
+
+/// Sets the given `WasiCxt` for the `Store`.
+///
+/// This will replace any existing `WasiCtx` associated with the `Store`.
+pub fn set_context(store: &Store, context: WasiCtx);
+```
+ 
+`add_to_engine` will add the various WASI functions to the given engine.
+
+The function implementations will expect that `Wasi::set_context` has been called on the `Store` prior to any invocations.
+
+If the context is not set, the WASI function implementations will trap.
+
+### WASI engine function example
+
+```rust
+let engine = Engine::default();
+Wasi::add_to_engine(&engine)?;
+ 
+let module = Module::new(
+    &engine,
+    r#"
+    (module
+        (import "wasi_snapshot_preview1" "fd_write" (func $fd_write (param i32 i32 i32 i32) (result i32)))
+        (memory (export "memory") 1)
+        (func (export "run")
+            i32.const 1
+            i32.const 0
+            i32.const 1
+            i32.const 12
+            call $fd_write
+            drop
+        )
+        (data (i32.const 0) "\0C\00\00\00\0D\00\00\00\00\00\00\00Hello world!\n")
+    )
+    "#,
+)?;
+
+// Set the WasiCtx in the store
+let store = Store::new(module.engine());
+Wasi::set_context(&store, WasiCtx::default()?);
```

Comment:
FWIW this doesn't look like it necessarily needs to be a function on`Wasi`, it could presumably just be `store.set_context(WasiCtx::default())`?

Reply:
The thinking behind this is that `Wasi` will likely want to wrap the `WasiCtx` in a `RefCell` and it is the only one that knows exactly what it expects the context type to be in the host functions.

Reply:
Ah ok that's a good point yeah, seems reasonable!